static/third/marked/lib/marked.js

```diff
@@ -519,14 +519,15 @@ inline.zulip = merge({}, inline.breaks, {
   stream: /^#\*\*([^\*]+)\*\*/,
   avatar: /^!avatar\(([^)]+)\)/,
   gravatar: /^!gravatar\(([^)]+)\)/,
-  tex: /^(\$\$([^ _$](\\\$|[^$])*)(?! )\$\$)\B/,
+  tex: /^(\$\$([^\n_$](\\\$|[^\n$])*)\$\$(?!\$))\B/,
   realm_filters: [],
   text: replace(inline.breaks.text)
     ('|', '|(\ud83c[\udd00-\udfff]|\ud83d[\udc00-\ude4f]|' +
           '\ud83d[\ude80-\udeff]|\ud83e[\udd00-\uddff]|' +
           '[\u2000-\u206F]|[\u2300-\u27BF]|[\u2B00-\u2BFF]|' +
           '[\u3000-\u303F]|[\u3200-\u32FF])|')
     (']|', '#@:]|')
+    ('^[', '^^\\${3,}|^^[')
```

Comment:
Can you explain this part of the change?

Reply:
Sure!  That's actually the part that tripped me up the most.
Previously, the code depended on there being no spaces after "$$" in order to avoid turning the following into an equation:

`Why have a little bit of money $$$ when you can have tons of money! $$$$$`

With the spaces constraint now being removed, we need to count our `$$`'s.
In the python markdown engine, this is easy.  We just attach a negative lookbehind before the first `$$` and a negative lookahead after the second `$$` in the regex that selects TeX, so we only match if there are exactly two `$`'s.  
The lookahead works just fine in marked.js, but the lookbehind doesn't.  Javascript does not have lookbehinds, and furthermore, even if it did, something like
`money money $$$$$ or maybe less money $$`
Would still match as TeX.  The reason for this is that the 'text:' regex in marked.js, without modification, will chip away one `$` at a time (since it always matches at least one character), until all that's left of the above string is
`$$ or maybe less money $$`
Which would clearly be matched by the `tex:` regex.  This is not desired behavior.  Hence, we need to consume substrings consisting of three or more `$` in the `text:` regex in marked.js, so the `$`'s belonging to, say `$$$$$$`, never start an equation.

Initially, I changed the initial `inline.text` regex on line 459 from
```js
text: /^[\s\S]+?(?=[\\<!\[_*`$]| {2,}\n|$)/
```
to
```js
text: /^\${3,}|^[\s\S]+?(?=[\\<!\[_*`$]| {2,}\n|$)/
```
so that it could match the $-triplets.   However, in the following lines, this regex gets modified in a few spots using the `replace( )` function, which ends up causing odd issues, like breaking some of the other replaces done in code before the change highlighted above.  So, I decided to leave the original `text:` regex alone and instead use the replace function down the road, which is what every other inline grammar modification seems to be doing.
` '^['` Is the beginning of the `text:` regex, and I replace it with `'^^\\${3,}|^^['`  The reason for the duplicated `^`'s is in the replace function's body itself:

```js
function replace(regex, opt) {
  regex = regex.source;
  opt = opt || '';
  return function self(name, val) {
    if (!name) return new RegExp(regex, opt);
    val = val.source || val;
    val = val.replace(/(^|[^\[])\^/g, '$1');
    regex = regex.replace(name, val);
    return self;
  };
}
```

That `val.replace()` turns the somewhat odd looking `^^\\${3,}|^^[` into `^\\${3,}|^[`, which is the desired substitution.

This whole mechanism using the replace function to add to the inline grammar seems quite hacky.  It really should have it's own function for adding things, IMO.  I didn't want to change too much in the file, though.  I think this is the most minimal and clean change that gets the job done.  I'm not entirely sure what the logic behind `replace()` chopping off `^`s from its arguments is, but it does.  To change that behavior would entail checking all of the numerous times the function is called throughout the file, and probably also entail refactoring a few of those calls, which might or might not be a good idea to do sometime!